File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans=nums[0]
        maxa,mina=ans,ans

        
        for i in range(1,len(nums)):
            if nums[i]<0:
                maxa,mina=mina,maxa
                
            maxa=max(nums[i],maxa*nums[i])
            mina=min(nums[i],mina*nums[i])
            ans=max(ans,maxa)

            
        return ans

File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int maxprod = Integer.MIN_VALUE;
        int nlen=nums.length;
        int prod = 1;
        for (int i = 0; i < nlen; i++) {
            prod *= nums[i];
            maxprod = Math.max(maxprod, prod);
            if (prod == 0)
                prod = 1;
        }
        prod = 1;
        for (int i = nlen - 1; i >= 0; i--) {
            prod *= nums[i];
            maxprod = Math.max(maxprod, prod);
            if (prod == 0)
                prod = 1;
        }
        return maxprod;
    }
}